metadata/src/utils/validation.rs

```diff
@@ -587,9 +578,9 @@ impl<'a> MetadataHasher<'a> {
     pub fn hash(&self) -> Hash {
         let metadata = self.metadata;
 
-        // Get the hashes of outer enums, considering only `specific_pallets` (if any are set).
+        // Get the hashes of outer enums.
         // If any of the typed that represent outer enums are encountered later, hashes from `top_level_enum_hashes` can be substituted.
-        let outer_enum_hashes = OuterEnumHashes::new(metadata, self.specific_pallets.as_deref());
+        let outer_enum_hashes = OuterEnumHashes::new(metadata);
 
```

Comment:
I'm still a bit confused by this.

Basically, the case I am thinking about is if we want to compare some "full" metadata with some "stripped down" metadata. We want to ensure that the remaining parts of the stripped metadata are in line with this full metadata (which would come form a node at runtime most probably).

To generate the same hash as the stripped metadata, we need to hash only the pallets that were kept around when hashing the full metadata.

To do this, we would previously generate `outer_enum_hashes` which hashed only the variants for those specific pallets (because we _always_ stripped them). Now, it looks like we _always_ hash every pallet in the full metadata, and so the hash would sometimes be different from the stripped metadata (ie whenever the stripped metadata actually removes variants, it will be different). 

I guess this isn't tested yet, since I don't think it should work as it stands (but maybe I'm missing something?)

I think this hashing function would have to also work out whether the variants would be stripped or not based on what we pass to `only_these_runtime_apis` and `only_these_pallets`? This would be a good reason to have a separate `find_type_id_in_return_pos` type function, so that we can lean on it in both places.

Reply:
This is essentially using the striping logic (without actually having to do the heavyweight stuff of removing types/pallets etc) in order that we can compare some stripped metadata that a user might have passed to the Subxt macro with some "full" metadata that you get from a node, to check that the two are compatible.

And so my concern with this bit of code is that it looks like the hash generated for some stripped metadata would be different from the hash gernerated from the same "full" metadata (assuming that the stripped metadata was able to strip variants from one or more outer enums).

Reply:
(happy to jump on a call or chat more about this if it doesn't make sense, and maybe I'm missing something too :))